library/CodeMonkeysRu/GCM/Response.php

```diff
@@ -162,11 +245,89 @@ function($result) {
                 return (
                     isset($result['error'])
                     &&
+                    (
                     ($result['error'] == "Unavailable")
+                    ||
+                    ($result['error'] == "InternalServerError")
+                    )
+                    );
+            });
+
+        return array_keys($filteredResults);
+    }
+    
+    public function someIdsUnavailable() {
+        $bool = (count($this->getUnavailableRegistrationIds()) != 0);
+        return $bool;
+    }
+
+    //invalid data key
+    
+    public function getInvalidDataKeysIds() {
+        if ($this->getFailureCount() == 0) {
+            return array();
+        }
+        $filteredResults = array_filter($this->results,
+            function($result) {
+                return (
+                    isset($result['error'])
+                    &&
+                    ($result['error'] == "InvalidDataKey")
                     );
             });
 
         return array_keys($filteredResults);
+    
     }
+    
+    public function existsInvalidDataKey() {
```

Comment:
All users are sent the same notification, so if one of them was sent an invalid data key, they all were. This is a major problem.